### PR TITLE
fixed validation error when using flash attention

### PR DESCRIPTION
**Pull request by francesco-bertolotti**
This PR addresses the issue https://github.com/pytorch/torchtitan/issues/2140

Briefly, the bug is related to the validation step.
The validate method did not pass the attention mask to the model.
When using flash attention, this leads to an error.

### CODE REVIEW

File: torchtitan/components/validate.py [hunk 1]
```diff
@@ -105,6 +113,13 @@ def validate(
             ):
                 break
 
+            # prepare attention mask
+            extra_kwargs : dict[str, Any] = {"attention_masks" : model_parts[0].get_attention_masks(
```
Comment (wwwjn) [inline torchtitan/components/validate.py]: For readability, can you make this into a function similar as: https://github.com/pytorch/torchtitan/blob/refs/heads/main/torchtitan/train.py#L416C9-L416C33

**Comment by francesco-bertolotti** [comment]
The new commit propagated the fix to the pipeline parallelism case. Further, I have encapsulated the creation of the attention mask in a method.

File: torchtitan/components/validate.py [hunk 2]
```diff
@@ -76,12 +78,43 @@ def __init__(
...
+        model_args.update_from_config(job_config)
+        self.model_args = model_args
+
         if self.job_config.validation.steps == -1:
             logger.warning(
                 "Setting validation steps to -1 might cause hangs because of "
                 "unequal sample counts across ranks when dataset is exhausted."
             )
 
+    def post_dataloading_process(
```
Comment (tianyu-l) [inline torchtitan/components/validate.py:92]: can we reuse the one in train.py?

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/components/validate.py (hunk 2 above)
Comment (francesco-bertolotti) [reply torchtitan/components/validate.py:92]: In principle, it would be possible, but I would need to make some modifications since their signatures differ. In particular, the trainer accesses `model_parts` through `self`.

One possible solution is to turn `post_dataloading_process` into a utility function that both the trainer and validator can call. However, this would mean you can no longer modify the behavior through inheritance—unless the current `post_dataloading_process` simply becomes a wrapper around the new utility function.|

If you have any suggestions, I can add a commit.

Comment (tianyu-l) [reply torchtitan/components/validate.py:92]: I see.
If you want to "modify the behavior", we probably should have a `PostDataloadingProcessor` class.
If not, I think it's fine to call a util directly in both trainer and validator.

WDYT?
cc @fegin

Comment (fegin) [reply torchtitan/components/validate.py:92]: I'm thinking this as well, I'm seeing the duplicated code in trainer and validator, `model_parts` is not a big problem, you can add it. See https://github.com/pytorch/torchtitan/pull/2144. But the duplicated logic is pretty concerning.

My initial proposal is to have an util function in `distributed.utils` so that both methods can call it.  The only worry is that I don't know whether `distributed.utils` is the best place to put it. It is not now because the util function purely unwrap the input but the util function will encapsulate the CP logic after https://github.com/pytorch/torchtitan/pull/2144. So `distributed.utils` seems to be a good place.


@francesco-bertolotti  you can do it as your PR may be landed first. I can rebase on top of your change.

Comment (tianyu-l) [reply torchtitan/components/validate.py:92]: agreed we should share code, but not sure about `distributed.utils`. Let's discuss offline.

**Review by tianyu-l — APPROVED** [review]
unblocking

**Comment by tianyu-l** [comment]
@francesco-bertolotti CPU unit test failed, could you take a look?

**Comment by francesco-bertolotti** [comment]
Hi @tianyu-l,

I looked into the error, and it turns out to be a circular import issue.

In `validate.py`, I need access to `model_args`. Since `Validator` doesn’t have direct access to it, I currently construct it via `get_train_spec`. However, importing `get_train_spec` triggers an import of `BaseValidator`, which lives in the same module as `Validator`, leading to the circular dependency.

I see two possible ways to resolve this:

1. Pass `model_args` directly to `Validator`, eliminating the need to import `get_train_spec`.
2. Move `BaseValidator` into its own module.

I’m leaning slightly toward changing the `Validator` signature, as it feels cleaner overall, though it would be a breaking change.

Curious to hear your thoughts.

**Comment by tianyu-l** [comment]
It seems this is being fixed in https://github.com/pytorch/torchtitan/pull/2144, with the first approach you mentioned.

Are you OK if we consolidate effort there?

**Comment by francesco-bertolotti** [comment]
Hi @tianyu-l , 

Since I do not think this PR is going to get merge can I close it?
